File: stage6-priority-inbox.ts
import axios from 'axios';

interface Notification {
  id: string;
  title: string;
  message: string;
  type: 'Placement' | 'Result' | 'Event';
  createdAt: string;
  isRead: boolean;
}

interface ScoredNotification extends Notification {
  score: number;
  ageHours: number;
  weight: number;
}

const TYPE_WEIGHTS: Record<string, number> = {
  'Placement': 100,
  'Result': 50,
  'Event': 10
};

function calculateAge(createdAt: string): number {
  const now = new Date();
  const created = new Date(createdAt);
  const diffMs = now.getTime() - created.getTime();
  return diffMs / (1000 * 60 * 60);
}

function calculateScore(notification: Notification): { score: number; weight: number } {
  const weight = TYPE_WEIGHTS[notification.type] || 0;
  const ageHours = calculateAge(notification.createdAt);
  
  const recencyFactor = Math.max(0, 1 - (ageHours / 168));
  const score = weight * recencyFactor;
  
  return { score, weight };
}

async function fetchNotificationsFromAPI(studentId: string): Promise<Notification[]> {
  try {
    const apiUrl = 'http://4.224.186.213/evaluation-service/notifications';
    
    const response = await axios.get(apiUrl, {
      params: {
        studentId: studentId
      },
      timeout: 5000
    });
    
    return response.data || [];
  } catch (error) {
    console.error('Error fetching notifications:', error);
    return [];
  }
}

function getTopNotifications(
  notifications: Notification[],
  limit: number = 10
): ScoredNotification[] {
  const unreadNotifications = notifications.filter(n => !n.isRead);
  
  const scoredNotifications: ScoredNotification[] = unreadNotifications.map(notif => {
    const { score, weight } = calculateScore(notif);
    return {
      ...notif,
      score,
      weight,
      ageHours: calculateAge(notif.createdAt)
    };
  });
  
  scoredNotifications.sort((a, b) => {
    if (b.score !== a.score) {
      return b.score - a.score;
    }
    return new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime();
  });
  
  return scoredNotifications.slice(0, limit);
}

function formatNotificationForDisplay(notification: ScoredNotification, index: number): string {
  const typeEmoji: Record<string, string> = {
    'Placement': '💼',
    'Result': '📊',
    'Event': '📢'
  };
  
  const emoji = typeEmoji[notification.type] || '📌';
  
  return `${index + 1}. ${emoji} [${notification.type}] ${notification.title}`;
}

async function displayPriorityInbox(studentId: string) {
  console.log('\n════════════════════════════════════════════════════════════');
  console.log('            PRIORITY INBOX - TOP 10 NOTIFICATIONS');
  console.log('════════════════════════════════════════════════════════════\n');
  
  const notifications = await fetchNotificationsFromAPI(studentId);
  const topNotifications = getTopNotifications(notifications, 10);
  
  if (topNotifications.length === 0) {
    console.log('No unread notifications found.');
    return;
  }
  
  console.log(`Student ID: ${studentId}`);
  console.log(`Total Unread: ${notifications.filter(n => !n.isRead).length}`);
  console.log(`Top Displaying: ${topNotifications.length}\n`);
  console.log('────────────────────────────────────────────────────────────\n');
  
  topNotifications.forEach((notif, index) => {
    console.log(formatNotificationForDisplay(notif, index));
    console.log(`   Type Weight: ${notif.weight} | Score: ${notif.score.toFixed(3)}`);
    console.log(`   Message: ${notif.message}`);
    console.log(`   Age: ${notif.ageHours.toFixed(1)} hours`);
    console.log(`   Created: ${notif.createdAt}`);
    console.log();
  });
  
  console.log('════════════════════════════════════════════════════════════');
}

class PriorityInboxManager {
  private priorityCache: Map<string, ScoredNotification[]> = new Map();
  private maxCacheSize: number = 10;
  private cacheExpiry: Map<string, number> = new Map();
  private cacheTTL: number = 3600000;
  
  async addNotification(studentId: string, notification: Notification): Promise<void> {
    const { score, weight } = calculateScore(notification);
    const scoredNotif: ScoredNotification = {
      ...notification,
      score,
      weight,
      ageHours: calculateAge(notification.createdAt)
    };
    
    const cacheKey = `student_${studentId}`;
    let topNotifications = this.priorityCache.get(cacheKey) || [];
    
    topNotifications.push(scoredNotif);
    topNotifications.sort((a, b) => b.score - a.score);
    topNotifications = topNotifications.slice(0, this.maxCacheSize);
    
    this.priorityCache.set(cacheKey, topNotifications);
    this.cacheExpiry.set(cacheKey, Date.now() + this.cacheTTL);
  }
  
  getTopNotifications(studentId: string, limit: number = 10): ScoredNotification[] {
    const cacheKey = `student_${studentId}`;
    const expiry = this.cacheExpiry.get(cacheKey) || 0;
    
    if (Date.now() > expiry) {
      this.priorityCache.delete(cacheKey);
      return [];
    }
    
    return (this.priorityCache.get(cacheKey) || []).slice(0, limit);
  }
  
  invalidateCache(studentId: string): void {
    const cacheKey = `student_${studentId}`;
    this.priorityCache.delete(cacheKey);
    this.cacheExpiry.delete(cacheKey);
  }
}

async function demonstratePriorityLogic() {
  console.log('\n════════════════════════════════════════════════════════════');
  console.log('        PRIORITY SCORING ALGORITHM DEMONSTRATION');
  console.log('════════════════════════════════════════════════════════════\n');
  
  const demoNotifications: Notification[] = [
    {
      id: '1',
      title: 'Placement Offer',
      message: 'Congratulations! You have received a placement offer.',
      type: 'Placement',
      createdAt: new Date(Date.now() - 1 * 60 * 60 * 1000).toISOString(),
      isRead: false
    },
    {
      id: '2',
      title: 'Semester Results',
      message: 'Your semester results are now available.',
      type: 'Result',
      createdAt: new Date(Date.now() - 5 * 60 * 60 * 1000).toISOString(),
      isRead: false
    },
    {
      id: '3',
      title: 'Campus Event',
      message: 'Upcoming tech symposium on June 25, 2026.',
      type: 'Event',
      createdAt: new Date(Date.now() - 10 * 60 * 60 * 1000).toISOString(),
      isRead: false
    },
    {
      id: '4',
      title: 'Interview Scheduled',
      message: 'Your interview is scheduled for June 22, 2026.',
      type: 'Placement',
      createdAt: new Date(Date.now() - 24 * 60 * 60 * 1000).toISOString(),
      isRead: false
    },
    {
      id: '5',
      title: 'Quiz Results Posted',
      message: 'Your quiz results have been posted.',
      type: 'Result',
      createdAt: new Date(Date.now() - 48 * 60 * 60 * 1000).toISOString(),
      isRead: false
    }
  ];
  
  const scored = demoNotifications.map(notif => {
    const { score, weight } = calculateScore(notif);
    return {
      ...notif,
      score,
      weight,
      ageHours: calculateAge(notif.createdAt)
    };
  });
  
  scored.sort((a, b) => b.score - a.score);
  
  console.log('Type Weights: Placement=100, Result=50, Event=10');
  console.log('Formula: Score = Weight × (1 - Age_Hours / 168)\n');
  console.log('────────────────────────────────────────────────────────────\n');
  
  scored.forEach((notif, index) => {
    console.log(`${index + 1}. ${notif.type}`);
    console.log(`   Title: ${notif.title}`);
    console.log(`   Weight: ${notif.weight}`);
    console.log(`   Age: ${notif.ageHours.toFixed(1)} hours`);
    console.log(`   Recency Factor: ${(1 - notif.ageHours / 168).toFixed(3)}`);
    console.log(`   FINAL SCORE: ${notif.score.toFixed(3)}`);
    console.log();
  });
}

async function main() {
  try {
    await demonstratePriorityLogic();
    
    const studentId = '1042';
    await displayPriorityInbox(studentId);
    
    console.log('\n✓ Priority Inbox implementation complete');
    console.log('✓ Features: Scoring, Sorting, Caching, Efficiency\n');
    
  } catch (error) {
    console.error('Error:', error);
  }
}

main();

File: stage6-priority-inbox.py
from datetime import datetime, timedelta, timezone
from typing import List, Dict, Tuple
import json

class Notification:
    def __init__(self, id: str, title: str, message: str, notif_type: str, created_at: str, is_read: bool):
        self.id = id
        self.title = title
        self.message = message
        self.type = notif_type
        self.created_at = created_at
        self.is_read = is_read

class ScoredNotification(Notification):
    def __init__(self, notification: Notification, score: float, weight: int, age_hours: float):
        super().__init__(notification.id, notification.title, notification.message, 
                        notification.type, notification.created_at, notification.is_read)
        self.score = score
        self.weight = weight
        self.age_hours = age_hours

TYPE_WEIGHTS = {
    'Placement': 100,
    'Result': 50,
    'Event': 10
}

def calculate_age_hours(created_at: str) -> float:
    now = datetime.now(timezone.utc)
    created = datetime.fromisoformat(created_at.replace('Z', '+00:00'))
    diff = now - created
    return diff.total_seconds() / 3600

def calculate_score(notification: Notification) -> Tuple[float, int]:
    weight = TYPE_WEIGHTS.get(notification.type, 0)
    age_hours = calculate_age_hours(notification.created_at)
    
    recency_factor = max(0, 1 - (age_hours / 168))
    score = weight * recency_factor
    
    return score, weight

def get_top_notifications(notifications: List[Notification], limit: int = 10) -> List[ScoredNotification]:
    unread = [n for n in notifications if not n.is_read]
    
    scored = []
    for notif in unread:
        score, weight = calculate_score(notif)
        age_hours = calculate_age_hours(notif.created_at)
        scored.append(ScoredNotification(notif, score, weight, age_hours))
    
    scored.sort(key=lambda x: (-x.score, datetime.fromisoformat(x.created_at.replace('Z', '+00:00'))))
    
    return scored[:limit]

def display_priority_inbox(notifications: List[Notification], student_id: str):
    print('\n' + '='*60)
    print('            PRIORITY INBOX - TOP 10 NOTIFICATIONS')
    print('='*60 + '\n')
    
    top_notifs = get_top_notifications(notifications, 10)
    
    if not top_notifs:
        print('No unread notifications found.')
        return
    
    unread_count = sum(1 for n in notifications if not n.is_read)
    
    print(f'Student ID: {student_id}')
    print(f'Total Unread: {unread_count}')
    print(f'Top Displaying: {len(top_notifs)}\n')
    print('-' * 60 + '\n')
    
    type_emoji = {
        'Placement': '💼',
        'Result': '📊',
        'Event': '📢'
    }
    
    for index, notif in enumerate(top_notifs, 1):
        emoji = type_emoji.get(notif.type, '📌')
        print(f'{index}. {emoji} [{notif.type}] {notif.title}')
        print(f'   Type Weight: {notif.weight} | Score: {notif.score:.3f}')
        print(f'   Message: {notif.message}')
        print(f'   Age: {notif.age_hours:.1f} hours')
        print(f'   Created: {notif.created_at}')
        print()
    
    print('='*60)

def demonstrate_priority_logic():
    print('\n' + '='*60)
    print('        PRIORITY SCORING ALGORITHM DEMONSTRATION')
    print('='*60 + '\n')
    
    now = datetime.now(timezone.utc)
    demo_notifications = [
        Notification('1', 'Placement Offer', 
                    'Congratulations! You have received a placement offer.',
                    'Placement',
                    (now - timedelta(hours=1)).isoformat().replace('+00:00', 'Z'), False),
        Notification('2', 'Semester Results',
                    'Your semester results are now available.',
                    'Result',
                    (now - timedelta(hours=5)).isoformat().replace('+00:00', 'Z'), False),
        Notification('3', 'Campus Event',
                    'Upcoming tech symposium on June 25, 2026.',
                    'Event',
                    (now - timedelta(hours=10)).isoformat().replace('+00:00', 'Z'), False),
        Notification('4', 'Interview Scheduled',
                    'Your interview is scheduled for June 22, 2026.',
                    'Placement',
                    (now - timedelta(hours=24)).isoformat().replace('+00:00', 'Z'), False),
        Notification('5', 'Quiz Results Posted',
                    'Your quiz results have been posted.',
                    'Result',
                    (now - timedelta(hours=48)).isoformat().replace('+00:00', 'Z'), False),
    ]
    
    print('Type Weights: Placement=100, Result=50, Event=10')
    print('Formula: Score = Weight × (1 - Age_Hours / 168)\n')
    print('-' * 60 + '\n')
    
    scored = []
    for notif in demo_notifications:
        score, weight = calculate_score(notif)
        age_hours = calculate_age_hours(notif.created_at)
        scored.append(ScoredNotification(notif, score, weight, age_hours))
    
    scored.sort(key=lambda x: (-x.score, datetime.fromisoformat(x.created_at.replace('Z', '+00:00'))))
    
    for index, notif in enumerate(scored, 1):
        recency_factor = 1 - (notif.age_hours / 168) if notif.age_hours < 168 else 0
        print(f'{index}. {notif.type}')
        print(f'   Title: {notif.title}')
        print(f'   Weight: {notif.weight}')
        print(f'   Age: {notif.age_hours:.1f} hours')
        print(f'   Recency Factor: {recency_factor:.3f}')
        print(f'   FINAL SCORE: {notif.score:.3f}')
        print()

class PriorityInboxManager:
    def __init__(self, max_cache_size: int = 10, cache_ttl_seconds: int = 3600):
        self.priority_cache = {}
        self.cache_expiry = {}
        self.max_cache_size = max_cache_size
        self.cache_ttl = cache_ttl_seconds
    
    def add_notification(self, student_id: str, notification: Notification):
        score, weight = calculate_score(notification)
        age_hours = calculate_age_hours(notification.created_at)
        scored_notif = ScoredNotification(notification, score, weight, age_hours)
        
        cache_key = f'student_{student_id}'
        top_notifs = self.priority_cache.get(cache_key, [])
        
        top_notifs.append(scored_notif)
        top_notifs.sort(key=lambda x: -x.score)
        top_notifs = top_notifs[:self.max_cache_size]
        
        self.priority_cache[cache_key] = top_notifs
        self.cache_expiry[cache_key] = datetime.now().timestamp() + self.cache_ttl
    
    def get_top_notifications(self, student_id: str, limit: int = 10) -> List[ScoredNotification]:
        cache_key = f'student_{student_id}'
        expiry = self.cache_expiry.get(cache_key, 0)
        
        if datetime.now().timestamp() > expiry:
            self.priority_cache.pop(cache_key, None)
            return []
        
        return self.priority_cache.get(cache_key, [])[:limit]
    
    def invalidate_cache(self, student_id: str):
        cache_key = f'student_{student_id}'
        self.priority_cache.pop(cache_key, None)
        self.cache_expiry.pop(cache_key, None)

def main():
    demonstrate_priority_logic()
    
    now = datetime.now(timezone.utc)
    demo_notifications = [
        Notification('1', 'Placement Offer', 
                    'Congratulations! You have received a placement offer.',
                    'Placement',
                    (now - timedelta(hours=1)).isoformat().replace('+00:00', 'Z'), False),
        Notification('2', 'Semester Results',
                    'Your semester results are now available.',
                    'Result',
                    (now - timedelta(hours=5)).isoformat().replace('+00:00', 'Z'), False),
        Notification('3', 'Campus Event',
                    'Upcoming tech symposium on June 25, 2026.',
                    'Event',
                    (now - timedelta(hours=10)).isoformat().replace('+00:00', 'Z'), False),
        Notification('4', 'Interview Scheduled',
                    'Your interview is scheduled for June 22, 2026.',
                    'Placement',
                    (now - timedelta(hours=24)).isoformat().replace('+00:00', 'Z'), False),
        Notification('5', 'Quiz Results Posted',
                    'Your quiz results have been posted.',
                    'Result',
                    (now - timedelta(hours=48)).isoformat().replace('+00:00', 'Z'), False),
    ]
    
    student_id = '1042'
    display_priority_inbox(demo_notifications, student_id)
    
    print('\n✓ Priority Inbox implementation complete')
    print('✓ Features: Scoring, Sorting, Caching, Efficiency\n')

if __name__ == '__main__':
    main()
